Source/Scene/ShadowMapShader.js

```diff
@@ -0,0 +1,368 @@
+    };
+
+    ShadowMapShader.createShadowReceiveVertexShader = function(vs, isTerrain, hasTerrainNormal) {
+        var defines = vs.defines.slice(0);
+        var sources = vs.sources.slice(0);
+
+        if (isTerrain) {
+            if (hasTerrainNormal) {
+                defines.push('GENERATE_POSITION_AND_NORMAL');
+            } else {
+                defines.push('GENERATE_POSITION');
+            }
+        }
+
+        return new ShaderSource({
+            defines : defines,
+            sources : sources
+        });
+    };
+
+    ShadowMapShader.createShadowReceiveFragmentShader = function(fs, shadowMap, castShadows, isTerrain, hasTerrainNormal) {
+        var normalVaryingName = findNormalVarying(fs);
+        var hasNormalVarying = (!isTerrain && defined(normalVaryingName)) || (isTerrain && hasTerrainNormal);
+
+        var positionVaryingName = findPositionVarying(fs);
+        var hasPositionVarying = defined(positionVaryingName);
+
+        var usesDepthTexture = shadowMap._usesDepthTexture;
+        var isPointLight = shadowMap._isPointLight;
+        var isSpotLight = shadowMap._isSpotLight;
+        var usesCubeMap = shadowMap._usesCubeMap;
+        var hasCascades = shadowMap._numberOfCascades > 1;
+        var cascadesEnabled = shadowMap._cascadesEnabled;
+        var debugVisualizeCascades = shadowMap.debugVisualizeCascades;
+        var softShadows = shadowMap.softShadows;
+        var bias = isPointLight ? shadowMap._pointBias : (isTerrain ? shadowMap._terrainBias : shadowMap._primitiveBias);
+        var exponentialShadows = shadowMap._exponentialShadows;
+
+        var defines = fs.defines.slice(0);
+        var sources = fs.sources.slice(0);
+
+        var length = sources.length;
+        for (var i = 0; i < length; ++i) {
+            sources[i] = ShaderSource.replaceMain(sources[i], 'czm_shadow_receive_main');
+        }
+
+        if (isPointLight && usesCubeMap) {
+            defines.push('USE_CUBE_MAP_SHADOW');
+        } else if (usesDepthTexture) {
+            defines.push('USE_SHADOW_DEPTH_TEXTURE');
+        }
+
+        if (softShadows && !isPointLight) {
+            defines.push('USE_SOFT_SHADOWS');
+        }
+
+        if (castShadows && bias.normalShading && hasNormalVarying) {
+            defines.push('USE_NORMAL_SHADING');
+            if (bias.normalShadingSmooth > 0.0) {
+                defines.push('USE_NORMAL_SHADING_SMOOTH');
+            }
+        }
+
+        if (exponentialShadows) {
+            defines.push('USE_EXPONENTIAL_SHADOW_MAPS');
+        }
+
+        var fsSource =
+            'uniform mat4 u_shadowMapMatrix; \n' +
+            'uniform vec3 u_shadowMapLightDirectionEC; \n' +
+            'uniform vec4 u_shadowMapLightPositionEC; \n' +
+            'uniform vec4 u_shadowMapNormalOffsetScaleDistanceMaxDistanceAndDarkness; \n' +
+            'uniform vec4 u_shadowMapTexelSizeDepthBiasAndNormalShadingSmooth; \n' +
+            'uniform sampler2D u_shadowMapTexture; \n' +
```

Comment:
Should we check `isPointLight` and then include just the right sampler so we're not relying on the compiler?